src/main/scala/io/iohk/ethereum/mining/BlockGenerator.scala

```diff
@@ -76,6 +64,32 @@ class BlockGenerator(blockchainStorages: BlockchainStorages, blockchainConfig: B
     result
   }
 
+  private def prepareTransactions(transactions: Seq[SignedTransaction], blockGasLimit: BigInt) = {
+    val sortedTransactions = transactions.groupBy(_.senderAddress).values.toList.flatMap { txsFromSender =>
+      val ordered = txsFromSender
+        .sortBy(-_.tx.gasPrice)
+        .sortBy(_.tx.nonce)
+        .scanLeft(None: Option[BigInt], None: Option[SignedTransaction]) { case ((prevNonce, _), stx) =>
```

Comment:
Even this is not wrong at all, maybe it would be more clear to replace the `scanLeft` and `flatMap` by a `foldLeft` with a condition whether to add or not at tx on last accumulated item. Something like (I haven't run it so it might not worn :disappointed: , read it as a reference impl )

```scala
.foldLeft(Seq.empty[SignedTransaction]) { (acc, stx) => {
   if(acc.lastOption.exists(_.nonce != stx.tx.nonce)) acc :+ stx
   else acc
}
```